In [1]:
from pathlib import Path

import numpy as np
import pandas as pd

# Project root
PROJECT_ROOT = Path.cwd().parent

# Input: existing engineered feature dataset
INPUT_FILE = PROJECT_ROOT / "data" / "processed" / "engineered_features.parquet"

# Output: new feature set
OUTPUT_FILE = PROJECT_ROOT / "data" / "processed" / "engineered_features_v2.parquet"

print("Project root :", PROJECT_ROOT)
print("Input file   :", INPUT_FILE)
print("Output file  :", OUTPUT_FILE)

assert INPUT_FILE.exists(), f"Input file not found: {INPUT_FILE}"

Project root : /Users/akashray/Lithology_Predictor
Input file   : /Users/akashray/Lithology_Predictor/data/processed/engineered_features.parquet
Output file  : /Users/akashray/Lithology_Predictor/data/processed/engineered_features_v2.parquet


In [2]:
df = pd.read_parquet(INPUT_FILE)

print("Shape:", df.shape)
print("\nColumns:")
print(df.columns.tolist())

print("\nFirst 5 rows:")
display(df.head())

print("\nData types:")
display(df.dtypes)

print("\nWells:", df["WELL"].nunique())

if "DEPT" in df.columns:
    print("\nDEPT is available.")
    print("Depth range:", df["DEPT"].min(), "to", df["DEPT"].max())
else:
    print("\nWARNING: DEPT is NOT available.")

Shape: (1430974, 46)

Columns:
['GR', 'RDEP', 'RMED', 'DTC', 'RHOB', 'FORCE_2020_LITHOFACIES_LITHOLOGY', 'WELL', 'RDEP_LOG10', 'RMED_LOG10', 'GR_MISSING', 'RDEP_LOG10_MISSING', 'RMED_LOG10_MISSING', 'DTC_MISSING', 'RHOB_MISSING', 'SPLIT', 'GR_ROLLMEAN_5', 'GR_ROLLSTD_5', 'GR_ROLLMEAN_9', 'GR_ROLLSTD_9', 'GR_ROLLMEAN_21', 'GR_ROLLSTD_21', 'RDEP_LOG10_ROLLMEAN_5', 'RDEP_LOG10_ROLLSTD_5', 'RDEP_LOG10_ROLLMEAN_9', 'RDEP_LOG10_ROLLSTD_9', 'RDEP_LOG10_ROLLMEAN_21', 'RDEP_LOG10_ROLLSTD_21', 'RMED_LOG10_ROLLMEAN_5', 'RMED_LOG10_ROLLSTD_5', 'RMED_LOG10_ROLLMEAN_9', 'RMED_LOG10_ROLLSTD_9', 'RMED_LOG10_ROLLMEAN_21', 'RMED_LOG10_ROLLSTD_21', 'DTC_ROLLMEAN_5', 'DTC_ROLLSTD_5', 'DTC_ROLLMEAN_9', 'DTC_ROLLSTD_9', 'DTC_ROLLMEAN_21', 'DTC_ROLLSTD_21', 'RHOB_ROLLMEAN_5', 'RHOB_ROLLSTD_5', 'RHOB_ROLLMEAN_9', 'RHOB_ROLLSTD_9', 'RHOB_ROLLMEAN_21', 'RHOB_ROLLSTD_21', 'RESISTIVITY_SEPARATION']

First 5 rows:


,GR,RDEP,RMED,DTC,RHOB,FORCE_2020_LITHOFACIES_LITHOLOGY,WELL,RDEP_LOG10,RMED_LOG10,GR_MISSING,...,DTC_ROLLSTD_9,DTC_ROLLMEAN_21,DTC_ROLLSTD_21,RHOB_ROLLMEAN_5,RHOB_ROLLSTD_5,RHOB_ROLLMEAN_9,RHOB_ROLLSTD_9,RHOB_ROLLMEAN_21,RHOB_ROLLSTD_21,RESISTIVITY_SEPARATION
0,80.200851,1.798681,1.611410,161.131180,1.884186,65000.0,15_9-13,0.254954,0.207206,0,...,0.435010,160.689362,0.770519,1.890167,0.006177,1.888490,0.006480,1.872489,0.020630,0.047748
1,79.262886,1.795641,1.618070,160.603470,1.889794,65000.0,15_9-13,0.254220,0.208997,0,...,0.406452,160.536517,0.905575,1.890604,0.005118,1.887023,0.006820,1.871460,0.019990,0.045222
2,74.821999,1.800733,1.626459,160.173615,1.896523,65000.0,15_9-13,0.255449,0.211243,0,...,0.493570,160.424934,0.955817,1.888490,0.006480,1.885838,0.006970,1.870925,0.019236,0.044206
3,72.878922,1.801517,1.621594,160.149429,1.891913,65000.0,15_9-13,0.255638,0.209942,0,...,0.755635,160.331406,0.982739,1.887590,0.007465,1.883588,0.009063,1.870178,0.018692,0.045696
4,71.729141,1.795299,1.602679,160.128342,1.880034,65000.0,15_9-13,0.254137,0.204847,0,...,0.761515,160.251672,0.996069,1.885378,0.008247,1.879549,0.014790,1.869431,0.018242,0.049290



Data types:


GR                                  float64
RDEP                                float64
RMED                                float64
DTC                                 float64
RHOB                                float64
FORCE_2020_LITHOFACIES_LITHOLOGY    float64
WELL                                    str
RDEP_LOG10                          float64
RMED_LOG10                          float64
GR_MISSING                             int8
RDEP_LOG10_MISSING                     int8
RMED_LOG10_MISSING                     int8
DTC_MISSING                            int8
RHOB_MISSING                           int8
SPLIT                                   str
GR_ROLLMEAN_5                       float64
GR_ROLLSTD_5                        float64
GR_ROLLMEAN_9                       float64
GR_ROLLSTD_9                        float64
GR_ROLLMEAN_21                      float64
GR_ROLLSTD_21                       float64
RDEP_LOG10_ROLLMEAN_5               float64
RDEP_LOG10_ROLLSTD_5            


Wells: 118



In [3]:
from pathlib import Path
import pandas as pd

PREPROCESSED_FILE = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "supervised_preprocessed.parquet"
)

print("File:", PREPROCESSED_FILE)
print("Exists:", PREPROCESSED_FILE.exists())

if PREPROCESSED_FILE.exists():
    pre_df = pd.read_parquet(PREPROCESSED_FILE)

    print("\nShape:", pre_df.shape)
    print("\nColumns:")
    print(pre_df.columns.tolist())

    depth_cols = [
        c for c in pre_df.columns
        if "DEPTH" in c.upper() or c.upper() in ["DEPT", "DEPTH"]
    ]

    print("\nPossible depth columns:", depth_cols)

File: /Users/akashray/Lithology_Predictor/data/processed/supervised_preprocessed.parquet
Exists: True

Shape: (1430974, 15)

Columns:
['GR', 'RDEP', 'RMED', 'DTC', 'RHOB', 'FORCE_2020_LITHOFACIES_LITHOLOGY', 'WELL', 'RDEP_LOG10', 'RMED_LOG10', 'GR_MISSING', 'RDEP_LOG10_MISSING', 'RMED_LOG10_MISSING', 'DTC_MISSING', 'RHOB_MISSING', 'SPLIT']

Possible depth columns: []


In [4]:
# Check the original FORCE 2020 LAS files
RAW_DIR = PROJECT_ROOT / "data" / "raw" / "FORCE_2020"

print("Raw directory:", RAW_DIR)
print("Exists:", RAW_DIR.exists())

if RAW_DIR.exists():
    las_files = sorted(RAW_DIR.glob("*.las"))

    print("Number of LAS files:", len(las_files))
    print("\nFirst 10 files:")
    for f in las_files[:10]:
        print(" ", f.name)

Raw directory: /Users/akashray/Lithology_Predictor/data/raw/FORCE_2020
Exists: True
Number of LAS files: 118

First 10 files:
  15_9-13.las
  15_9-14.las
  15_9-15.las
  15_9-17.las
  15_9-23.las
  16_1-2.las
  16_1-6 A.las
  16_10-1.las
  16_10-2.las
  16_10-3.las


In [ ]:
import lasio
import numpy as np
import pandas as pd

RAW_DIR = PROJECT_ROOT / "data" / "raw" / "FORCE_2020"

MODEL_CURVES = ["GR", "RDEP", "RMED", "DTC", "RHOB"]
TARGET_COL = "FORCE_2020_LITHOFACIES_LITHOLOGY"

depth_parts = []
count_check = []

for i, las_path in enumerate(sorted(RAW_DIR.glob("*.las")), start=1):

    las = lasio.read(las_path)

    # ---------------------------------------------------------
    # Read raw LAS data
    # ---------------------------------------------------------
    raw = las.df().reset_index()

    # Normalize column names
    raw.columns = [str(c).upper() for c in raw.columns]

    if "DEPT" not in raw.columns:
        raise ValueError(f"DEPT not found in {las_path.name}")

    # ---------------------------------------------------------
    # Handle missing RMED exactly as preprocessing did
    # ---------------------------------------------------------
    if "RMED" not in raw.columns:
        raw["RMED"] = np.nan

    # Check remaining required columns
    required = [
        "DEPT",
        "GR",
        "RDEP",
        "RMED",
        "DTC",
        "RHOB",
        TARGET_COL,
    ]

    missing = [c for c in required if c not in raw.columns]

    if missing:
        raise ValueError(
            f"{las_path.name}: missing required columns: {missing}"
        )

    # ---------------------------------------------------------
    # Convert LAS NULL values to NaN
    # ---------------------------------------------------------
    raw = raw.replace([-999.25, -999.250000], np.nan)

    # ---------------------------------------------------------
    # SAME cleaning rules as established preprocessing
    # ---------------------------------------------------------

    # GR
    raw.loc[raw["GR"] < 0, "GR"] = np.nan

    # Resistivity
    raw.loc[raw["RDEP"] <= 0, "RDEP"] = np.nan
    raw.loc[raw["RMED"] <= 0, "RMED"] = np.nan

    # Sonic
    raw.loc[raw["DTC"] <= 0, "DTC"] = np.nan

    # Density
    raw.loc[raw["RHOB"] <= 0, "RHOB"] = np.nan

    # ---------------------------------------------------------
    # Same row-selection logic
    # ---------------------------------------------------------

    has_any_feature = raw[MODEL_CURVES].notna().any(axis=1)
    has_label = raw[TARGET_COL].notna()

    keep = has_any_feature & has_label

    selected = raw.loc[
        keep,
        ["DEPT"] + MODEL_CURVES + [TARGET_COL]
    ].copy()

    # ---------------------------------------------------------
    # Recover exact well name
    # ---------------------------------------------------------
    well_name = las_path.stem.replace("_", "/", 1)

    selected["WELL"] = well_name

    depth_parts.append(
        selected[["WELL", "DEPT"]]
    )

    count_check.append({
        "WELL": well_name,
        "raw_rows": len(raw),
        "selected_rows": len(selected),
        "RMED_available": "RMED" in las.df().columns,
    })

    if i % 20 == 0 or i == 118:
        print(f"Processed {i}/118 wells")


# -------------------------------------------------------------
# Combine all wells
# -------------------------------------------------------------

depth_recovered = pd.concat(
    depth_parts,
    ignore_index=True
)

count_check = pd.DataFrame(count_check)


# -------------------------------------------------------------
# Summary
# -------------------------------------------------------------

print("\n==============================================")
print("DEPTH RECONSTRUCTION SUMMARY")
print("==============================================")

print("Recovered rows :", len(depth_recovered))
print("Processed rows :", len(df))

print(
    "Row difference :",
    len(depth_recovered) - len(df)
)

print(
    "\nRecovered wells:",
    depth_recovered["WELL"].nunique()
)

print(
    "Processed wells:",
    df["WELL"].nunique()
)

print("\nFirst recovered rows:")
display(depth_recovered.head())

print("\nRMED availability:")
display(
    count_check["RMED_available"].value_counts()
)

print("\nPer-well count check:")
display(count_check.head)

Processed 20/118 wells
Processed 40/118 wells
Processed 60/118 wells
Processed 80/118 wells
Processed 100/118 wells
Processed 118/118 wells

DEPTH RECONSTRUCTION SUMMARY
Recovered rows : 1430974
Processed rows : 1430974
Row difference : 0

Recovered wells: 118
Processed wells: 118

First recovered rows:


,WELL,DEPT
0,15/9-13,494.528
1,15/9-13,494.680
2,15/9-13,494.832
3,15/9-13,494.984
4,15/9-13,495.136



RMED availability:


RMED_available
True     117
False      1
Name: count, dtype: int64


Per-well count check:


,WELL,raw_rows,selected_rows,RMED_available
0,15/9-13,21441,18270,True
1,15/9-14,22875,20336,True
2,15/9-15,20954,17717,True
3,15/9-17,19879,17351,True
4,15/9-23,20494,11063,True
5,16/1-2,18547,1734,True
6,16/1-6 A,13513,3623,True
7,16/10-1,20147,17675,True
8,16/10-2,20131,2437,True
9,16/10-3,17995,15956,True


In [11]:
# ============================================================
# CELL 6 — ALIGN RECOVERED DEPTH WITH PROCESSED DATA
# ============================================================

# Our processed dataset uses the LAS filename convention:
# 15_9-13, 15_9-14, etc.
#
# The recovered dataframe currently contains:
# 15/9-13, 15/9-14, etc.
#
# Convert the recovered names back to the processed convention.

depth_recovered["WELL"] = (
    depth_recovered["WELL"]
    .astype(str)
    .str.replace("/", "_", regex=False)
)

# ------------------------------------------------------------
# Basic checks
# ------------------------------------------------------------

print("Processed wells :", df["WELL"].nunique())
print("Recovered wells :", depth_recovered["WELL"].nunique())

print("\nProcessed first wells:")
print(df["WELL"].drop_duplicates().head(10).tolist())

print("\nRecovered first wells:")
print(depth_recovered["WELL"].drop_duplicates().head(10).tolist())


# ------------------------------------------------------------
# Check whether the WELL sequence matches row-by-row
# ------------------------------------------------------------

processed_wells = df["WELL"].reset_index(drop=True)
recovered_wells = depth_recovered["WELL"].reset_index(drop=True)

well_alignment = processed_wells.equals(recovered_wells)

print("\n==============================================")
print("WELL ALIGNMENT CHECK")
print("==============================================")
print("Row count:", len(df))
print("WELL sequence matches:", well_alignment)


# ------------------------------------------------------------
# If it does not match, show the first mismatch
# ------------------------------------------------------------

if not well_alignment:

    mismatch = np.flatnonzero(
        processed_wells.to_numpy() != recovered_wells.to_numpy()
    )

    print("\nNumber of mismatched rows:", len(mismatch))

    if len(mismatch) > 0:
        idx = mismatch[0]

        print("\nFirst mismatch:")
        print("Row index:", idx)
        print("Processed WELL:", processed_wells.iloc[idx])
        print("Recovered WELL:", recovered_wells.iloc[idx])

        print("\nProcessed context:")
        display(
            df.iloc[max(0, idx - 3):idx + 4][["WELL"]]
        )

        print("\nRecovered context:")
        display(
            depth_recovered.iloc[max(0, idx - 3):idx + 4][["WELL", "DEPT"]]
        )

        raise ValueError(
            "WELL sequence still does not align. "
            "Do not calculate gradients yet."
        )

print("\n✓ WELL sequence alignment passed.")

Processed wells : 118
Recovered wells : 118

Processed first wells:
['15_9-13', '15_9-14', '15_9-15', '15_9-17', '15_9-23', '16_1-2', '16_1-6 A', '16_10-1', '16_10-2', '16_10-3']

Recovered first wells:
['15_9-13', '15_9-14', '15_9-15', '15_9-17', '15_9-23', '16_1-2', '16_1-6 A', '16_10-1', '16_10-2', '16_10-3']

WELL ALIGNMENT CHECK
Row count: 1430974
WELL sequence matches: True

✓ WELL sequence alignment passed.


In [12]:
# ============================================================
# CELL 7 — ATTACH VERIFIED DEPTH TO ENGINEERED FEATURES
# ============================================================

# Make a copy so the original engineered dataset remains untouched.
df_v2 = df.copy()

# Add the recovered depth.
df_v2["DEPT"] = depth_recovered["DEPT"].to_numpy()

# ------------------------------------------------------------
# Check the resulting dataset
# ------------------------------------------------------------

print("Original engineered shape :", df.shape)
print("New feature-set shape    :", df_v2.shape)

print("\nDEPT missing values:")
print(df_v2["DEPT"].isna().sum())

print("\nDEPT data type:")
print(df_v2["DEPT"].dtype)

# ------------------------------------------------------------
# Check depth range for the whole dataset
# ------------------------------------------------------------

print("\nOverall depth range:")
print("Minimum:", df_v2["DEPT"].min())
print("Maximum:", df_v2["DEPT"].max())

# ------------------------------------------------------------
# Check depth ordering within each well
# ------------------------------------------------------------

depth_diff = (
    df_v2
    .groupby("WELL")["DEPT"]
    .diff()
)

print("\nDepth difference statistics:")
display(depth_diff.describe())

# ------------------------------------------------------------
# Check for problematic depth differences
# ------------------------------------------------------------

print("\nDepth differences <= 0:")
print((depth_diff <= 0).sum())

print("\nDepth differences > 0:")
print((depth_diff > 0).sum())

# ------------------------------------------------------------
# Show a few wells
# ------------------------------------------------------------

print("\nExample depth values:")
display(
    df_v2[["WELL", "DEPT"]]
    .head(20)
)

Original engineered shape : (1430974, 46)
New feature-set shape    : (1430974, 47)

DEPT missing values:
0

DEPT data type:
float64

Overall depth range:
Minimum: 136.086
Maximum: 5436.632

Depth difference statistics:


count    1.430856e+06
mean     1.530733e-01
std      2.971792e-01
min      1.520000e-01
25%      1.520000e-01
50%      1.520000e-01
75%      1.520000e-01
max      2.693440e+02
Name: DEPT, dtype: float64


Depth differences <= 0:
0

Depth differences > 0:
1430856

Example depth values:


,WELL,DEPT
0,15_9-13,494.528
1,15_9-13,494.680
2,15_9-13,494.832
3,15_9-13,494.984
4,15_9-13,495.136
5,15_9-13,495.288
6,15_9-13,495.440
7,15_9-13,495.592
8,15_9-13,495.744
9,15_9-13,495.896


In [13]:
# ============================================================
# CELL 8 — INSPECT DEPTH GAPS
# ============================================================

# Calculate depth differences within each well.
depth_gap = (
    df_v2
    .groupby("WELL")["DEPT"]
    .diff()
)

# Remove the first row of every well (NaN difference).
depth_gap = depth_gap.dropna()

print("==============================================")
print("DEPTH GAP ANALYSIS")
print("==============================================")

print("\nAll depth-gap statistics:")
display(depth_gap.describe())

# ------------------------------------------------------------
# Count different gap sizes
# ------------------------------------------------------------

print("\nGap <= 0.152 m:")
print((depth_gap <= 0.152001).sum())

print("\nGap > 0.152 m:")
print((depth_gap > 0.152001).sum())

print("\nGap > 0.304 m:")
print((depth_gap > 0.304001).sum())

print("\nGap > 1 m:")
print((depth_gap > 1.0).sum())

print("\nGap > 10 m:")
print((depth_gap > 10.0).sum())

print("\nGap > 100 m:")
print((depth_gap > 100.0).sum())

# ------------------------------------------------------------
# Show the largest gaps
# ------------------------------------------------------------

print("\nLargest 20 depth gaps:")

largest_gaps = (
    df_v2[["WELL", "DEPT"]]
    .copy()
)

largest_gaps["DEPTH_GAP"] = (
    largest_gaps
    .groupby("WELL")["DEPT"]
    .diff()
)

display(
    largest_gaps
    .sort_values("DEPTH_GAP", ascending=False)
    .head(20)
)

DEPTH GAP ANALYSIS

All depth-gap statistics:


count    1.430856e+06
mean     1.530733e-01
std      2.971792e-01
min      1.520000e-01
25%      1.520000e-01
50%      1.520000e-01
75%      1.520000e-01
max      2.693440e+02
Name: DEPT, dtype: float64


Gap <= 0.152 m:
1430684

Gap > 0.152 m:
172

Gap > 0.304 m:
170

Gap > 1 m:
137

Gap > 10 m:
32

Gap > 100 m:
3

Largest 20 depth gaps:


,WELL,DEPT,DEPTH_GAP
1315885,35_4-1,2954.164000,269.344
1314063,35_4-1,2408.028000,112.632
977186,34_10-35,3192.847000,111.872
1017648,34_12-1,4116.493584,83.296
224834,16_7-5,2393.439985,83.144
1308247,35_4-1,1409.084000,48.032
1408564,35_9-8,3149.301600,39.368
1047805,34_3-2 S,2541.238800,37.696
1390595,35_9-6 S,3215.512000,30.400
1048065,34_3-2 S,2610.854800,30.248


In [19]:
# ============================================================
# CELL 9 — CALCULATE FULL-DEPTH GRADIENTS
# ============================================================

# These are the log curves for which we want depth gradients.
GRADIENT_CURVES = [
    "GR",
    "RDEP_LOG10",
    "RMED_LOG10",
    "DTC",
    "RHOB",
]

# Get the LAS files once.
las_files = sorted(RAW_DIR.glob("*.las"))

# This list will hold the gradient dataframe for each well.
gradient_parts = []

print("Calculating gradients from the full depth sequences...\n")


# ============================================================
# PROCESS EACH WELL
# ============================================================

for i, las_path in enumerate(las_files, start=1):

    # --------------------------------------------------------
    # 1. Read original LAS file
    # --------------------------------------------------------
    las = lasio.read(las_path)

    raw = las.df().reset_index()

    # Normalize all curve names to uppercase.
    raw.columns = [str(c).upper() for c in raw.columns]

    # --------------------------------------------------------
    # 2. Handle missing RMED
    # --------------------------------------------------------
    if "RMED" not in raw.columns:
        raw["RMED"] = np.nan

    # --------------------------------------------------------
    # 3. Replace LAS NULL values with NaN
    # --------------------------------------------------------
    raw = raw.replace([-999.25, -999.250000], np.nan)

    # --------------------------------------------------------
    # 4. Apply our established cleaning rules
    # --------------------------------------------------------

    # GR <= 0 is not valid for our preprocessing.
    raw.loc[raw["GR"] < 0, "GR"] = np.nan

    # Resistivity must be positive.
    raw.loc[raw["RDEP"] <= 0, "RDEP"] = np.nan
    raw.loc[raw["RMED"] <= 0, "RMED"] = np.nan

    # Sonic and density must be positive.
    raw.loc[raw["DTC"] <= 0, "DTC"] = np.nan
    raw.loc[raw["RHOB"] <= 0, "RHOB"] = np.nan

    # --------------------------------------------------------
    # 5. Create log10 resistivity
    # --------------------------------------------------------

    raw["RDEP_LOG10"] = np.log10(raw["RDEP"])
    raw["RMED_LOG10"] = np.log10(raw["RMED"])

    # --------------------------------------------------------
    # 6. Calculate gradients using the FULL depth sequence
    # --------------------------------------------------------

    depth = raw["DEPT"].to_numpy(dtype=float)

    gradient_data = {
        "DEPT": depth
    }

    for curve in GRADIENT_CURVES:

        values = raw[curve].to_numpy(dtype=float)

        # Start with all NaN.
        gradient = np.full(len(raw), np.nan, dtype=float)

        # Difference between consecutive depth samples.
        dz = depth[1:] - depth[:-1]

        # Difference between consecutive log values.
        dx = values[1:] - values[:-1]

        # Only calculate where both values and depth are valid.
        valid = (
            np.isfinite(values[:-1])
            & np.isfinite(values[1:])
            & np.isfinite(depth[:-1])
            & np.isfinite(depth[1:])
            & (dz > 0)
        )

        # Forward gradient:
        #
        # dX/dz = (X[i+1] - X[i]) / (z[i+1] - z[i])
        #
        gradient[:-1][valid] = dx[valid] / dz[valid]

        gradient_data[f"{curve}_GRAD"] = gradient

    # --------------------------------------------------------
    # 7. Recover the same WELL name used by our processed data
    # --------------------------------------------------------

    well_name = las_path.stem

    gradient_well = pd.DataFrame(gradient_data)

    gradient_well["WELL"] = well_name

    # Put columns in a predictable order.
    gradient_well = gradient_well[
        ["WELL", "DEPT"]
        + [f"{curve}_GRAD" for curve in GRADIENT_CURVES]
    ]

    gradient_parts.append(gradient_well)

    if i % 20 == 0 or i == len(las_files):
        print(f"Processed {i}/{len(las_files)} wells")


# ============================================================
# COMBINE ALL WELLS
# ============================================================

full_gradients = pd.concat(
    gradient_parts,
    ignore_index=True
)


# ============================================================
# SUMMARY
# ============================================================

print("\n==============================================")
print("FULL GRADIENT DATASET")
print("==============================================")

print("Rows :", len(full_gradients))
print("Wells:", full_gradients["WELL"].nunique())

print("\nColumns:")
print(full_gradients.columns.tolist())


# ============================================================
# GRADIENT MISSINGNESS
# ============================================================

gradient_columns = [
    f"{curve}_GRAD"
    for curve in GRADIENT_CURVES
]

print("\nGradient missingness (%):")

gradient_missingness = (
    full_gradients[gradient_columns]
    .isna()
    .mean()
    .mul(100)
    .round(3)
)

display(gradient_missingness)


# ============================================================
# GRADIENT STATISTICS
# ============================================================

print("\nGradient statistics:")

gradient_statistics = (
    full_gradients[gradient_columns]
    .describe()
    .T
)

display(gradient_statistics)

Calculating gradients from the full depth sequences...

Processed 20/118 wells
Processed 40/118 wells
Processed 60/118 wells
Processed 80/118 wells
Processed 100/118 wells
Processed 118/118 wells

FULL GRADIENT DATASET
Rows : 2337639
Wells: 118

Columns:
['WELL', 'DEPT', 'GR_GRAD', 'RDEP_LOG10_GRAD', 'RMED_LOG10_GRAD', 'DTC_GRAD', 'RHOB_GRAD']

Gradient missingness (%):


GR_GRAD             1.814
RDEP_LOG10_GRAD     5.074
RMED_LOG10_GRAD    10.489
DTC_GRAD           23.661
RHOB_GRAD          41.206
dtype: float64


Gradient statistics:


,count,mean,std,min,25%,50%,75%,max
GR_GRAD,2295243.0,0.013194,23.431477,-1909.094961,-9.224641,0.009587,9.275938,1766.524666
RDEP_LOG10_GRAD,2219036.0,0.000143,0.235752,-22.006264,-0.035848,-0.000078,0.035566,25.044544
RMED_LOG10_GRAD,2092435.0,0.000196,0.287358,-22.296477,-0.046296,-0.000005,0.046062,22.338632
DTC_GRAD,1784531.0,-0.024477,13.340870,-487.992136,-3.731602,-0.002510,3.659299,848.337977
RHOB_GRAD,1374397.0,0.000214,0.180852,-8.706394,-0.049356,0.000322,0.050258,6.637915


In [20]:
# ============================================================
# CELL 10 — VALIDATE GRADIENT DEPTH SPACING
# ============================================================

# ------------------------------------------------------------
# 1. Calculate depth difference inside each well
# ------------------------------------------------------------

full_gradients["DZ"] = (
    full_gradients
    .groupby("WELL")["DEPT"]
    .diff()
)

print("==============================================")
print("FULL-DEPTH GRADIENT SPACING")
print("==============================================")

print("\nDepth-gap statistics:")
display(
    full_gradients["DZ"]
    .dropna()
    .describe()
)


# ------------------------------------------------------------
# 2. Count normal and abnormal depth steps
# ------------------------------------------------------------

normal_step = np.isclose(
    full_gradients["DZ"],
    0.152,
    atol=1e-6
)

abnormal_step = (
    full_gradients["DZ"].notna()
    & ~normal_step
)

print("\nNormal 0.152 m transitions:")
print(normal_step.sum())

print("\nAbnormal transitions:")
print(abnormal_step.sum())

print(
    "\nPercentage abnormal:",
    round(abnormal_step.mean() * 100, 5),
    "%"
)


# ------------------------------------------------------------
# 3. Show the largest abnormal gaps
# ------------------------------------------------------------

print("\nLargest depth gaps in full LAS sequences:")

display(
    full_gradients[
        ["WELL", "DEPT", "DZ"]
    ]
    .sort_values("DZ", ascending=False)
    .head(20)
)


# ------------------------------------------------------------
# 4. Invalidate gradients across abnormal depth gaps
# ------------------------------------------------------------

gradient_columns = [
    "GR_GRAD",
    "RDEP_LOG10_GRAD",
    "RMED_LOG10_GRAD",
    "DTC_GRAD",
    "RHOB_GRAD",
]

for column in gradient_columns:
    full_gradients.loc[abnormal_step, column] = np.nan


# ------------------------------------------------------------
# 5. Recalculate statistics after removing
#    gradients across abnormal depth gaps
# ------------------------------------------------------------

print("\n==============================================")
print("CLEANED GRADIENT STATISTICS")
print("==============================================")

gradient_statistics_clean = (
    full_gradients[gradient_columns]
    .describe()
    .T
)

display(gradient_statistics_clean)


# ------------------------------------------------------------
# 6. Missingness after depth-gap filtering
# ------------------------------------------------------------

print("\nGradient missingness after depth-gap filtering (%):")

gradient_missingness_clean = (
    full_gradients[gradient_columns]
    .isna()
    .mean()
    .mul(100)
    .round(3)
)

display(gradient_missingness_clean)

FULL-DEPTH GRADIENT SPACING

Depth-gap statistics:


count    2.337521e+06
mean     1.520000e-01
std      1.544250e-10
min      1.520000e-01
25%      1.520000e-01
50%      1.520000e-01
75%      1.520000e-01
max      1.520001e-01
Name: DZ, dtype: float64


Normal 0.152 m transitions:
2337521

Abnormal transitions:
0

Percentage abnormal: 0.0 %

Largest depth gaps in full LAS sequences:


,WELL,DEPT,DZ
1297719,33_9-17,1000.017985,0.152
27370,15_9-14,1000.012001,0.152
627299,25_4-5,1000.134805,0.152
2306107,7_1-1,1000.072800,0.152
1624150,34_4-10 R,1000.122467,0.152
1318092,34_10-16 R,1000.064008,0.152
840001,26_4-1,1000.088199,0.152
2208357,35_9-5,1000.055587,0.152
1543293,34_3-1 A,1000.136715,0.152
2048110,35_3-7 S,1000.139851,0.152



CLEANED GRADIENT STATISTICS


,count,mean,std,min,25%,50%,75%,max
GR_GRAD,2295243.0,0.013194,23.431477,-1909.094961,-9.224641,0.009587,9.275938,1766.524666
RDEP_LOG10_GRAD,2219036.0,0.000143,0.235752,-22.006264,-0.035848,-0.000078,0.035566,25.044544
RMED_LOG10_GRAD,2092435.0,0.000196,0.287358,-22.296477,-0.046296,-0.000005,0.046062,22.338632
DTC_GRAD,1784531.0,-0.024477,13.340870,-487.992136,-3.731602,-0.002510,3.659299,848.337977
RHOB_GRAD,1374397.0,0.000214,0.180852,-8.706394,-0.049356,0.000322,0.050258,6.637915



Gradient missingness after depth-gap filtering (%):


GR_GRAD             1.814
RDEP_LOG10_GRAD     5.074
RMED_LOG10_GRAD    10.489
DTC_GRAD           23.661
RHOB_GRAD          41.206
dtype: float64

In [21]:
# ============================================================
# CELL 11 — MERGE VERIFIED GRADIENTS INTO ML DATASET
# ============================================================

gradient_columns = [
    "GR_GRAD",
    "RDEP_LOG10_GRAD",
    "RMED_LOG10_GRAD",
    "DTC_GRAD",
    "RHOB_GRAD",
]

# ------------------------------------------------------------
# 1. Check that WELL + DEPT uniquely identifies a sample
# ------------------------------------------------------------

duplicate_gradient_keys = full_gradients.duplicated(
    subset=["WELL", "DEPT"]
).sum()

duplicate_ml_keys = df_v2.duplicated(
    subset=["WELL", "DEPT"]
).sum()

print("Duplicate (WELL, DEPT) keys in full gradients:",
      duplicate_gradient_keys)

print("Duplicate (WELL, DEPT) keys in ML dataset:",
      duplicate_ml_keys)

if duplicate_gradient_keys > 0:
    raise ValueError(
        "Duplicate WELL + DEPT found in full gradient dataset."
    )

if duplicate_ml_keys > 0:
    raise ValueError(
        "Duplicate WELL + DEPT found in ML dataset."
    )


# ------------------------------------------------------------
# 2. Keep only the gradient information we need
# ------------------------------------------------------------

gradient_lookup = full_gradients[
    ["WELL", "DEPT"] + gradient_columns
].copy()


# ------------------------------------------------------------
# 3. Merge gradients into our existing feature dataset
# ------------------------------------------------------------

df_v2_gradient = df_v2.merge(
    gradient_lookup,
    on=["WELL", "DEPT"],
    how="left",
    validate="one_to_one"
)


# ------------------------------------------------------------
# 4. Verify row count
# ------------------------------------------------------------

print("\n==============================================")
print("GRADIENT MERGE CHECK")
print("==============================================")

print("Original ML rows :", len(df_v2))
print("New V2 rows      :", len(df_v2_gradient))

if len(df_v2_gradient) != len(df_v2):
    raise ValueError(
        "Row count changed after gradient merge."
    )

print("✓ Row count preserved.")


# ------------------------------------------------------------
# 5. Check that every ML row received a gradient record
# ------------------------------------------------------------

print("\nGradient missingness after merge (%):")

gradient_missingness_v2 = (
    df_v2_gradient[gradient_columns]
    .isna()
    .mean()
    .mul(100)
    .round(3)
)

display(gradient_missingness_v2)


# ------------------------------------------------------------
# 6. Show final shape and columns
# ------------------------------------------------------------

print("\nOriginal feature count:", df.shape[1])

print(
    "Feature count after adding DEPT:",
    df_v2.shape[1]
)

print(
    "Feature count after adding gradients:",
    df_v2_gradient.shape[1]
)

print("\nNew gradient columns:")
print(gradient_columns)


# ------------------------------------------------------------
# 7. Inspect a few rows
# ------------------------------------------------------------

print("\nExample rows with gradients:")

display(
    df_v2_gradient[
        [
            "WELL",
            "DEPT",
            "GR",
            "DTC",
            "RHOB",
            "GR_GRAD",
            "DTC_GRAD",
            "RHOB_GRAD",
        ]
    ].head(15)
)

Duplicate (WELL, DEPT) keys in full gradients: 0
Duplicate (WELL, DEPT) keys in ML dataset: 0

GRADIENT MERGE CHECK
Original ML rows : 1430974
New V2 rows      : 1430974
✓ Row count preserved.

Gradient missingness after merge (%):


GR_GRAD             0.092
RDEP_LOG10_GRAD     0.776
RMED_LOG10_GRAD     3.470
DTC_GRAD            6.056
RHOB_GRAD          13.144
dtype: float64


Original feature count: 46
Feature count after adding DEPT: 47
Feature count after adding gradients: 52

New gradient columns:
['GR_GRAD', 'RDEP_LOG10_GRAD', 'RMED_LOG10_GRAD', 'DTC_GRAD', 'RHOB_GRAD']

Example rows with gradients:


,WELL,DEPT,GR,DTC,RHOB,GR_GRAD,DTC_GRAD,RHOB_GRAD
0,15_9-13,494.528,80.200851,161.131180,1.884186,-6.170825,-3.471776,0.036893
1,15_9-13,494.680,79.262886,160.603470,1.889794,-29.216365,-2.827996,0.044271
2,15_9-13,494.832,74.821999,160.173615,1.896523,-12.783402,-0.159113,-0.030329
3,15_9-13,494.984,72.878922,160.149429,1.891913,-7.564344,-0.138735,-0.078150
4,15_9-13,495.136,71.729141,160.128342,1.880034,1.876831,0.137831,-0.002281
5,15_9-13,495.288,72.014420,160.149292,1.879687,3.774141,7.244010,-0.006294
6,15_9-13,495.440,72.588089,161.250381,1.878731,-8.585779,6.342336,-0.071670
7,15_9-13,495.592,71.283051,162.214416,1.867837,-10.273783,-4.206005,-0.135553
8,15_9-13,495.744,69.721436,161.575104,1.847233,-20.024400,-6.524638,-0.071865
9,15_9-13,495.896,66.677727,160.583359,1.836309,-5.163996,-6.309007,0.058028


In [22]:
# ============================================================
# CELL 12 — SAVE FEATURE SET V2
# ============================================================

V2_FILE = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "engineered_features_v2.parquet"
)

# ------------------------------------------------------------
# 1. Save the complete V2 dataset
# ------------------------------------------------------------

df_v2_gradient.to_parquet(
    V2_FILE,
    index=False
)

print("Saved Feature Set V2:")
print(V2_FILE)


# ------------------------------------------------------------
# 2. Reload the saved file
# ------------------------------------------------------------

v2_check = pd.read_parquet(V2_FILE)


# ------------------------------------------------------------
# 3. Verify the saved dataset
# ------------------------------------------------------------

print("\n==============================================")
print("FEATURE SET V2 VERIFICATION")
print("==============================================")

print("Original V2 shape :", df_v2_gradient.shape)
print("Reloaded V2 shape :", v2_check.shape)

print(
    "\nShape matches:",
    df_v2_gradient.shape == v2_check.shape
)

print(
    "Gradient columns present:",
    all(
        column in v2_check.columns
        for column in gradient_columns
    )
)

print(
    "DEPT present:",
    "DEPT" in v2_check.columns
)

print(
    "Wells:",
    v2_check["WELL"].nunique()
)

print(
    "Rows:",
    len(v2_check)
)


# ------------------------------------------------------------
# 4. Final V2 column summary
# ------------------------------------------------------------

print("\nNew gradient columns:")
print(gradient_columns)

print("\nFeature Set V2 columns:")
print(v2_check.columns.tolist())

Saved Feature Set V2:
/Users/akashray/Lithology_Predictor/data/processed/engineered_features_v2.parquet

FEATURE SET V2 VERIFICATION
Original V2 shape : (1430974, 52)
Reloaded V2 shape : (1430974, 52)

Shape matches: True
Gradient columns present: True
DEPT present: True
Wells: 118
Rows: 1430974

New gradient columns:
['GR_GRAD', 'RDEP_LOG10_GRAD', 'RMED_LOG10_GRAD', 'DTC_GRAD', 'RHOB_GRAD']

Feature Set V2 columns:
['GR', 'RDEP', 'RMED', 'DTC', 'RHOB', 'FORCE_2020_LITHOFACIES_LITHOLOGY', 'WELL', 'RDEP_LOG10', 'RMED_LOG10', 'GR_MISSING', 'RDEP_LOG10_MISSING', 'RMED_LOG10_MISSING', 'DTC_MISSING', 'RHOB_MISSING', 'SPLIT', 'GR_ROLLMEAN_5', 'GR_ROLLSTD_5', 'GR_ROLLMEAN_9', 'GR_ROLLSTD_9', 'GR_ROLLMEAN_21', 'GR_ROLLSTD_21', 'RDEP_LOG10_ROLLMEAN_5', 'RDEP_LOG10_ROLLSTD_5', 'RDEP_LOG10_ROLLMEAN_9', 'RDEP_LOG10_ROLLSTD_9', 'RDEP_LOG10_ROLLMEAN_21', 'RDEP_LOG10_ROLLSTD_21', 'RMED_LOG10_ROLLMEAN_5', 'RMED_LOG10_ROLLSTD_5', 'RMED_LOG10_ROLLMEAN_9', 'RMED_LOG10_ROLLSTD_9', 'RMED_LOG10_ROLLMEAN_21'

now the controlled experiment on the newly designed feature

In [23]:
# ============================================================
# CELL 13 — PREPARE V2 FOR MACHINE LEARNING
# ============================================================

# Reload the saved Feature Set V2.
v2 = pd.read_parquet(V2_FILE)

TARGET = "FORCE_2020_LITHOFACIES_LITHOLOGY"

# ------------------------------------------------------------
# 1. Define the original W5 feature set
# ------------------------------------------------------------

base_features = [
    "GR",
    "RDEP_LOG10",
    "RMED_LOG10",
    "DTC",
    "RHOB",

    "GR_MISSING",
    "RDEP_LOG10_MISSING",
    "RMED_LOG10_MISSING",
    "DTC_MISSING",
    "RHOB_MISSING",

    "GR_ROLLMEAN_5",
    "GR_ROLLSTD_5",
    "GR_ROLLMEAN_9",
    "GR_ROLLSTD_9",
    "GR_ROLLMEAN_21",
    "GR_ROLLSTD_21",

    "RDEP_LOG10_ROLLMEAN_5",
    "RDEP_LOG10_ROLLSTD_5",
    "RDEP_LOG10_ROLLMEAN_9",
    "RDEP_LOG10_ROLLSTD_9",
    "RDEP_LOG10_ROLLMEAN_21",
    "RDEP_LOG10_ROLLSTD_21",

    "RMED_LOG10_ROLLMEAN_5",
    "RMED_LOG10_ROLLSTD_5",
    "RMED_LOG10_ROLLMEAN_9",
    "RMED_LOG10_ROLLSTD_9",
    "RMED_LOG10_ROLLMEAN_21",
    "RMED_LOG10_ROLLSTD_21",

    "DTC_ROLLMEAN_5",
    "DTC_ROLLSTD_5",
    "DTC_ROLLMEAN_9",
    "DTC_ROLLSTD_9",
    "DTC_ROLLMEAN_21",
    "DTC_ROLLSTD_21",

    "RHOB_ROLLMEAN_5",
    "RHOB_ROLLSTD_5",
    "RHOB_ROLLMEAN_9",
    "RHOB_ROLLSTD_9",
    "RHOB_ROLLMEAN_21",
    "RHOB_ROLLSTD_21",

    "RESISTIVITY_SEPARATION",
]

# ------------------------------------------------------------
# 2. Define the new gradient features
# ------------------------------------------------------------

gradient_features = [
    "GR_GRAD",
    "RDEP_LOG10_GRAD",
    "RMED_LOG10_GRAD",
    "DTC_GRAD",
    "RHOB_GRAD",
]

# ------------------------------------------------------------
# 3. Combine them
# ------------------------------------------------------------

v2_features = base_features + gradient_features

print("Base W5 features     :", len(base_features))
print("Gradient features    :", len(gradient_features))
print("Total V2 ML features :", len(v2_features))

assert len(base_features) == 41
assert len(gradient_features) == 5
assert len(v2_features) == 46

# ------------------------------------------------------------
# 4. Verify every feature exists
# ------------------------------------------------------------

missing_features = [
    feature
    for feature in v2_features
    if feature not in v2.columns
]

if missing_features:
    raise ValueError(
        f"Missing V2 features: {missing_features}"
    )

print("\n✓ All 46 V2 features are present.")


# ------------------------------------------------------------
# 5. Check infinite values
# ------------------------------------------------------------

inf_counts = np.isinf(
    v2[v2_features]
).sum()

print("\nInfinite values:")
display(
    inf_counts[inf_counts > 0]
)

if inf_counts.sum() == 0:
    print("✓ No infinite values found.")


# ------------------------------------------------------------
# 6. Define ML objects
# ------------------------------------------------------------

X_v2 = v2[v2_features].copy()

y_v2 = v2[TARGET].astype(int)

wells_v2 = v2["WELL"].copy()

print("\n==============================================")
print("V2 ML DATASET")
print("==============================================")

print("X shape :", X_v2.shape)
print("y shape :", y_v2.shape)
print("Wells   :", wells_v2.nunique())

print("\nTarget distribution:")
display(
    y_v2.value_counts().sort_index()
)

Base W5 features     : 41
Gradient features    : 5
Total V2 ML features : 46

✓ All 46 V2 features are present.

Infinite values:


Series([], dtype: int64)

✓ No infinite values found.

V2 ML DATASET
X shape : (1430974, 46)
y shape : (1430974,)
Wells   : 118

Target distribution:


FORCE_2020_LITHOFACIES_LITHOLOGY
30000    207671
65000    877042
65030    180446
70000     69498
70032     14043
74000      2391
80000     41038
86000      1807
88000     14712
90000      4754
93000       141
99000     17431
Name: count, dtype: int64

In [24]:
# ============================================================
# CELL 14 — FREEZE DEVELOPMENT WELL SPLIT
# ============================================================

from sklearn.model_selection import train_test_split

MANIFEST_FILE = (
    PROJECT_ROOT
    / "data"
    / "splits"
    / "force2020_split.csv"
)

print("Manifest:", MANIFEST_FILE)
print("Exists  :", MANIFEST_FILE.exists())

if not MANIFEST_FILE.exists():
    raise FileNotFoundError(
        f"Official split manifest not found: {MANIFEST_FILE}"
    )


# ------------------------------------------------------------
# 1. Load official FORCE split
# ------------------------------------------------------------

manifest = pd.read_csv(MANIFEST_FILE)

print("\nManifest columns:")
print(manifest.columns.tolist())

print("\nManifest shape:")
print(manifest.shape)

display(manifest.head())

Manifest: /Users/akashray/Lithology_Predictor/data/splits/force2020_split.csv
Exists  : True

Manifest columns:
['WELL', 'SPLIT']

Manifest shape:
(118, 2)


,WELL,SPLIT
0,15_9-13,TRAIN
1,15_9-14,PUBLIC
2,15_9-15,TRAIN
3,15_9-17,TRAIN
4,15_9-23,BLIND_TEST


In [25]:
# ============================================================
# CELL 15 — FREEZE DEVELOPMENT SPLIT
# ============================================================

from sklearn.model_selection import train_test_split

# ------------------------------------------------------------
# Official groups
# ------------------------------------------------------------

official_train_wells = (
    manifest.loc[
        manifest["SPLIT"].eq("TRAIN"),
        "WELL"
    ]
    .astype(str)
    .tolist()
)

official_public_wells = (
    manifest.loc[
        manifest["SPLIT"].eq("PUBLIC"),
        "WELL"
    ]
    .astype(str)
    .tolist()
)

official_blind_wells = (
    manifest.loc[
        manifest["SPLIT"].eq("BLIND_TEST"),
        "WELL"
    ]
    .astype(str)
    .tolist()
)


# ------------------------------------------------------------
# Verify official split first
# ------------------------------------------------------------

print("Official split:")
print(f"TRAIN      : {len(official_train_wells)} wells")
print(f"PUBLIC     : {len(official_public_wells)} wells")
print(f"BLIND_TEST : {len(official_blind_wells)} wells")


if len(official_train_wells) != 98:
    raise ValueError(
        f"Expected 98 official TRAIN wells, "
        f"found {len(official_train_wells)}"
    )

if len(official_public_wells) != 10:
    raise ValueError(
        f"Expected 10 PUBLIC wells, "
        f"found {len(official_public_wells)}"
    )

if len(official_blind_wells) != 10:
    raise ValueError(
        f"Expected 10 BLIND_TEST wells, "
        f"found {len(official_blind_wells)}"
    )


# ------------------------------------------------------------
# Split ONLY the official TRAIN wells
# ------------------------------------------------------------

dev_train_wells, dev_valid_wells = train_test_split(
    official_train_wells,
    test_size=20,
    random_state=42,
    shuffle=True,
)


# ------------------------------------------------------------
# Sort for reproducibility / readability
# ------------------------------------------------------------

dev_train_wells = sorted(dev_train_wells)
dev_valid_wells = sorted(dev_valid_wells)


# ------------------------------------------------------------
# Safety checks
# ------------------------------------------------------------

assert len(dev_train_wells) == 78
assert len(dev_valid_wells) == 20

assert set(dev_train_wells).isdisjoint(
    set(dev_valid_wells)
)

assert (
    set(dev_train_wells) | set(dev_valid_wells)
) == set(official_train_wells)

assert set(dev_train_wells).isdisjoint(
    set(official_public_wells)
)

assert set(dev_valid_wells).isdisjoint(
    set(official_public_wells)
)

assert set(dev_train_wells).isdisjoint(
    set(official_blind_wells)
)

assert set(dev_valid_wells).isdisjoint(
    set(official_blind_wells)
)


# ------------------------------------------------------------
# Print results
# ------------------------------------------------------------

print("\n" + "=" * 60)
print("FROZEN DEVELOPMENT SPLIT")
print("=" * 60)

print(f"\nDEV_TRAIN      : {len(dev_train_wells)} wells")
print(f"DEV_VALIDATION : {len(dev_valid_wells)} wells")

print("\nDEV_TRAIN wells:")
print(dev_train_wells)

print("\nDEV_VALIDATION wells:")
print(dev_valid_wells)


# ------------------------------------------------------------
# Save development manifest
# ------------------------------------------------------------

dev_rows = []

for well in dev_train_wells:
    dev_rows.append({
        "WELL": well,
        "DEV_SPLIT": "DEV_TRAIN",
        "OFFICIAL_SPLIT": "TRAIN",
    })

for well in dev_valid_wells:
    dev_rows.append({
        "WELL": well,
        "DEV_SPLIT": "DEV_VALIDATION",
        "OFFICIAL_SPLIT": "TRAIN",
    })

development_manifest = pd.DataFrame(dev_rows)

DEV_MANIFEST_FILE = (
    PROJECT_ROOT
    / "data"
    / "splits"
    / "development_split.csv"
)

development_manifest.to_csv(
    DEV_MANIFEST_FILE,
    index=False
)

print(
    f"\n✓ Development split saved to:\n"
    f"  {DEV_MANIFEST_FILE}"
)

print("\nFinal structure:")
print("  DEV_TRAIN      :", len(dev_train_wells))
print("  DEV_VALIDATION :", len(dev_valid_wells))
print("  PUBLIC         :", len(official_public_wells))
print("  BLIND_TEST     :", len(official_blind_wells))

Official split:
TRAIN      : 98 wells
PUBLIC     : 10 wells
BLIND_TEST : 10 wells

FROZEN DEVELOPMENT SPLIT

DEV_TRAIN      : 78 wells
DEV_VALIDATION : 20 wells

DEV_TRAIN wells:
['15_9-15', '15_9-17', '16_1-2', '16_10-1', '16_10-2', '16_10-3', '16_10-5', '16_11-1 ST3', '16_2-16', '16_4-1', '16_5-3', '16_7-4', '16_7-5', '16_8-1', '25_11-15', '25_11-19 S', '25_11-5', '25_2-14', '25_2-7', '25_3-1', '25_5-1', '25_5-4', '25_6-1', '25_6-2', '25_7-2', '25_8-5 S', '25_8-7', '25_9-1', '26_4-1', '29_6-1', '30_3-3', '30_3-5 S', '31_2-1', '31_2-7', '31_2-9', '31_3-1', '31_3-3', '31_3-4', '31_4-10', '31_4-5', '31_5-4 S', '31_6-5', '31_6-8', '32_2-1', '33_5-2', '33_6-3 S', '33_9-1', '33_9-17', '34_10-19', '34_10-21', '34_10-35', '34_11-2 S', '34_12-1', '34_2-4', '34_3-1 A', '34_4-10 R', '34_5-1 A', '34_5-1 S', '34_7-13', '34_7-21', '34_8-3', '34_8-7 R', '35_11-1', '35_11-10', '35_11-11', '35_11-13', '35_11-6', '35_11-7', '35_12-1', '35_3-7 S', '35_4-1', '35_8-4', '35_9-10 S', '35_9-2', '35_9-5', '3

In [26]:
# ============================================================
# CELL 16 — AUDIT DEVELOPMENT SPLIT
# ============================================================

TARGET = "FORCE_2020_LITHOFACIES_LITHOLOGY"

# ------------------------------------------------------------
# 1. Create a WELL → development split lookup
# ------------------------------------------------------------

dev_split_lookup = {}

for well in dev_train_wells:
    dev_split_lookup[well] = "DEV_TRAIN"

for well in dev_valid_wells:
    dev_split_lookup[well] = "DEV_VALIDATION"


# ------------------------------------------------------------
# 2. Assign every V2 row to its development split
# ------------------------------------------------------------

v2_dev = v2[
    v2["WELL"].isin(
        set(dev_train_wells) | set(dev_valid_wells)
    )
].copy()

v2_dev["DEV_SPLIT"] = v2_dev["WELL"].map(
    dev_split_lookup
)


# ------------------------------------------------------------
# 3. Basic checks
# ------------------------------------------------------------

print("==============================================")
print("DEVELOPMENT DATASET")
print("==============================================")

print("Rows:", len(v2_dev))
print("Wells:", v2_dev["WELL"].nunique())

print("\nRows by development split:")
display(
    v2_dev["DEV_SPLIT"].value_counts()
)

print("\nWells by development split:")
display(
    v2_dev.groupby("DEV_SPLIT")["WELL"]
    .nunique()
)


# ------------------------------------------------------------
# 4. Lithology counts
# ------------------------------------------------------------

class_counts = (
    v2_dev
    .groupby(["DEV_SPLIT", TARGET])
    .size()
    .unstack(fill_value=0)
)

print("\n==============================================")
print("LITHOLOGY SAMPLE COUNTS")
print("==============================================")

display(class_counts)


# ------------------------------------------------------------
# 5. Lithology percentages within each split
# ------------------------------------------------------------

class_percent = (
    class_counts
    .div(class_counts.sum(axis=1), axis=0)
    .mul(100)
    .round(3)
)

print("\n==============================================")
print("LITHOLOGY PERCENTAGE")
print("==============================================")

display(class_percent)


# ------------------------------------------------------------
# 6. Number of wells containing each lithology
# ------------------------------------------------------------

well_presence = (
    v2_dev
    .groupby([TARGET, "DEV_SPLIT"])["WELL"]
    .nunique()
    .unstack(fill_value=0)
)

print("\n==============================================")
print("WELLS CONTAINING EACH LITHOLOGY")
print("==============================================")

display(well_presence)


# ------------------------------------------------------------
# 7. Identify classes absent from validation
# ------------------------------------------------------------

all_classes = sorted(
    v2_dev[TARGET].unique()
)

validation_counts = (
    class_counts
    .loc["DEV_VALIDATION"]
    .reindex(all_classes, fill_value=0)
)

missing_in_validation = validation_counts[
    validation_counts == 0
].index.tolist()

print("\n==============================================")
print("VALIDATION RARE-CLASS CHECK")
print("==============================================")

if missing_in_validation:
    print(
        "WARNING — classes absent from DEV_VALIDATION:"
    )
    print(missing_in_validation)
else:
    print(
        "✓ Every lithology class is represented "
        "in DEV_VALIDATION."
    )

DEVELOPMENT DATASET
Rows: 1171557
Wells: 98

Rows by development split:


DEV_SPLIT
DEV_TRAIN         894965
DEV_VALIDATION    276592
Name: count, dtype: int64


Wells by development split:


DEV_SPLIT
DEV_TRAIN         78
DEV_VALIDATION    20
Name: WELL, dtype: int64


LITHOLOGY SAMPLE COUNTS


FORCE_2020_LITHOFACIES_LITHOLOGY,30000.0,65000.0,65030.0,70000.0,70032.0,74000.0,80000.0,86000.0,88000.0,90000.0,93000.0,99000.0
DEV_SPLIT,,,,,,,,,,,,
DEV_TRAIN,131392,544884,121266,43889,7938,1193,22389,845,4315,3306,141,13407
DEV_VALIDATION,38013,176301,29339,12431,2575,495,10947,240,3899,514,0,1838



LITHOLOGY PERCENTAGE


FORCE_2020_LITHOFACIES_LITHOLOGY,30000.0,65000.0,65030.0,70000.0,70032.0,74000.0,80000.0,86000.0,88000.0,90000.0,93000.0,99000.0
DEV_SPLIT,,,,,,,,,,,,
DEV_TRAIN,14.681,60.883,13.550,4.904,0.887,0.133,2.502,0.094,0.482,0.369,0.016,1.498
DEV_VALIDATION,13.743,63.740,10.607,4.494,0.931,0.179,3.958,0.087,1.410,0.186,0.000,0.665



WELLS CONTAINING EACH LITHOLOGY


DEV_SPLIT,DEV_TRAIN,DEV_VALIDATION
FORCE_2020_LITHOFACIES_LITHOLOGY,,
30000.0,78,20
65000.0,78,20
65030.0,77,20
70000.0,75,19
70032.0,10,1
74000.0,29,8
80000.0,59,14
86000.0,4,2
88000.0,2,1



VALIDATION RARE-CLASS CHECK
WARNING — classes absent from DEV_VALIDATION:
[93000.0]


In [27]:
# ============================================================
# CELL 17 — BUILD TRAIN / VALIDATION MATRICES
# ============================================================

TARGET = "FORCE_2020_LITHOFACIES_LITHOLOGY"

# ------------------------------------------------------------
# 1. Reload the verified V2 dataset
# ------------------------------------------------------------

v2 = pd.read_parquet(V2_FILE)

print("V2 shape:", v2.shape)


# ------------------------------------------------------------
# 2. Select only development wells
# ------------------------------------------------------------

dev_wells = set(dev_train_wells) | set(dev_valid_wells)

dev_data = v2[
    v2["WELL"].isin(dev_wells)
].copy()

print("Development rows:", len(dev_data))
print("Development wells:", dev_data["WELL"].nunique())


# ------------------------------------------------------------
# 3. Create train / validation masks
# ------------------------------------------------------------

train_mask = dev_data["WELL"].isin(
    set(dev_train_wells)
)

valid_mask = dev_data["WELL"].isin(
    set(dev_valid_wells)
)

print("\nRows:")
print("DEV_TRAIN      :", train_mask.sum())
print("DEV_VALIDATION :", valid_mask.sum())


# ------------------------------------------------------------
# 4. Build X and y
# ------------------------------------------------------------

X_train_v2 = dev_data.loc[
    train_mask,
    v2_features
].copy()

y_train_v2 = dev_data.loc[
    train_mask,
    TARGET
].astype(int)

X_valid_v2 = dev_data.loc[
    valid_mask,
    v2_features
].copy()

y_valid_v2 = dev_data.loc[
    valid_mask,
    TARGET
].astype(int)


# ------------------------------------------------------------
# 5. Check dimensions
# ------------------------------------------------------------

print("\n==============================================")
print("V2 TRAIN / VALIDATION MATRICES")
print("==============================================")

print("X_train :", X_train_v2.shape)
print("y_train :", y_train_v2.shape)

print("X_valid :", X_valid_v2.shape)
print("y_valid :", y_valid_v2.shape)


# ------------------------------------------------------------
# 6. Calculate class weights ONLY from DEV_TRAIN
# ------------------------------------------------------------

train_class_counts = (
    y_train_v2
    .value_counts()
    .sort_index()
)

n_train = len(y_train_v2)
n_classes = len(train_class_counts)

class_weights_v2 = np.sqrt(
    n_train /
    (n_classes * train_class_counts)
)

print("\n==============================================")
print("DEV_TRAIN CLASS WEIGHTS")
print("==============================================")

weight_table = pd.DataFrame({
    "CLASS": train_class_counts.index,
    "SAMPLES": train_class_counts.values,
    "WEIGHT": class_weights_v2.values
})

display(weight_table)


# ------------------------------------------------------------
# 7. Convert class weights → sample weights
# ------------------------------------------------------------

sample_weights_v2 = (
    y_train_v2
    .map(class_weights_v2)
    .to_numpy()
)

print("\nSample-weight statistics:")

print(
    "Minimum:",
    sample_weights_v2.min()
)

print(
    "Maximum:",
    sample_weights_v2.max()
)

print(
    "Mean:",
    sample_weights_v2.mean()
)

print(
    "Number of samples:",
    len(sample_weights_v2)
)


# ------------------------------------------------------------
# 8. Final safety checks
# ------------------------------------------------------------

assert len(X_train_v2) == len(y_train_v2)
assert len(X_valid_v2) == len(y_valid_v2)

assert len(sample_weights_v2) == len(y_train_v2)

assert set(
    dev_train_wells
).isdisjoint(
    set(dev_valid_wells)
)

print("\n✓ Training/validation construction passed.")
print("✓ Class weights calculated from DEV_TRAIN only.")
print("✓ Validation data has not been used for weighting.")

V2 shape: (1430974, 52)
Development rows: 1171557
Development wells: 98

Rows:
DEV_TRAIN      : 894965
DEV_VALIDATION : 276592

V2 TRAIN / VALIDATION MATRICES
X_train : (894965, 46)
y_train : (894965,)
X_valid : (276592, 46)
y_valid : (276592,)

DEV_TRAIN CLASS WEIGHTS


,CLASS,SAMPLES,WEIGHT
0,30000,131392,0.753404
1,65000,544884,0.369965
2,65030,121266,0.784229
3,70000,43889,1.303571
4,70032,7938,3.065186
5,74000,1193,7.906644
6,80000,22389,1.825135
7,86000,845,9.394724
8,88000,4315,4.157402
9,90000,3306,4.749643



Sample-weight statistics:
Minimum: 0.36996477403109324
Maximum: 22.998676598785163
Mean: 0.6748453290692437
Number of samples: 894965

✓ Training/validation construction passed.
✓ Class weights calculated from DEV_TRAIN only.
✓ Validation data has not been used for weighting.


In [28]:
# ============================================================
# CELL 18 — G1: W5 + GRADIENTS
# ============================================================

from sklearn.ensemble import ExtraTreesClassifier
from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    f1_score,
    classification_report,
    confusion_matrix,
)

import time


# ------------------------------------------------------------
# 1. Create the G1 model
# ------------------------------------------------------------

g1_model = ExtraTreesClassifier(
    n_estimators=100,
    max_depth=20,
    min_samples_leaf=5,
    class_weight=None,
    n_jobs=-1,
    random_state=42,
)

print("G1 model:")
print(g1_model)


# ------------------------------------------------------------
# 2. Train using DEV_TRAIN only
# ------------------------------------------------------------

print("\nTraining G1...")

train_start = time.perf_counter()

g1_model.fit(
    X_train_v2,
    y_train_v2,
    sample_weight=sample_weights_v2,
)

train_time = time.perf_counter() - train_start

print(
    f"Training completed in {train_time:.2f} seconds"
)


# ------------------------------------------------------------
# 3. Predict DEV_VALIDATION
# ------------------------------------------------------------

print("\nPredicting DEV_VALIDATION...")

prediction_start = time.perf_counter()

y_pred_g1 = g1_model.predict(X_valid_v2)

prediction_time = time.perf_counter() - prediction_start

print(
    f"Prediction completed in {prediction_time:.2f} seconds"
)


# ------------------------------------------------------------
# 4. Calculate main metrics
# ------------------------------------------------------------

accuracy_g1 = accuracy_score(
    y_valid_v2,
    y_pred_g1
)

balanced_accuracy_g1 = balanced_accuracy_score(
    y_valid_v2,
    y_pred_g1
)

macro_f1_g1 = f1_score(
    y_valid_v2,
    y_pred_g1,
    average="macro",
    zero_division=0,
)

weighted_f1_g1 = f1_score(
    y_valid_v2,
    y_pred_g1,
    average="weighted",
    zero_division=0,
)


# ------------------------------------------------------------
# 5. Print results
# ------------------------------------------------------------

print("\n" + "=" * 60)
print("G1 — W5 + GRADIENTS")
print("=" * 60)

print(f"Accuracy          : {accuracy_g1:.6f}")
print(f"Balanced Accuracy : {balanced_accuracy_g1:.6f}")
print(f"Macro F1          : {macro_f1_g1:.6f}")
print(f"Weighted F1       : {weighted_f1_g1:.6f}")

print(f"\nTraining time     : {train_time:.2f} s")
print(f"Prediction time   : {prediction_time:.2f} s")


# ------------------------------------------------------------
# 6. Detailed classification report
# ------------------------------------------------------------

print("\nClassification report:")

g1_report = classification_report(
    y_valid_v2,
    y_pred_g1,
    labels=sorted(y_valid_v2.unique()),
    zero_division=0,
)

print(g1_report)

G1 model:
ExtraTreesClassifier(max_depth=20, min_samples_leaf=5, n_jobs=-1,
                     random_state=42)

Training G1...
Training completed in 26.97 seconds

Predicting DEV_VALIDATION...
Prediction completed in 1.53 seconds

G1 — W5 + GRADIENTS
Accuracy          : 0.720122
Balanced Accuracy : 0.400922
Macro F1          : 0.302777
Weighted F1       : 0.700995

Training time     : 26.97 s
Prediction time   : 1.53 s

Classification report:
              precision    recall  f1-score   support

       30000       0.61      0.65      0.63     38013
       65000       0.82      0.89      0.85    176301
       65030       0.31      0.26      0.28     29339
       70000       0.60      0.62      0.61     12431
       70032       0.00      0.00      0.00      2575
       74000       0.18      0.02      0.04       495
       80000       0.65      0.17      0.27     10947
       86000       0.06      0.81      0.11       240
       88000       0.00      0.00      0.00      3899
       90

/Users/akashray/Lithology_Predictor/.venv/lib/python3.13/site-packages/sklearn/metrics/_classification.py:2939: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")


In [29]:
# ============================================================
# CELL 19 — G2: XGBOOST + GRADIENTS
# ============================================================

import time
import numpy as np

try:
    from xgboost import XGBClassifier
    print("✓ XGBoost is available.")
except ImportError as e:
    raise ImportError(
        "XGBoost is not installed in this environment."
    ) from e


# ------------------------------------------------------------
# 1. Encode lithology labels as 0 ... 11
# ------------------------------------------------------------

classes = np.sort(y_train_v2.unique())

class_to_index = {
    cls: i
    for i, cls in enumerate(classes)
}

index_to_class = {
    i: cls
    for cls, i in class_to_index.items()
}

y_train_xgb = y_train_v2.map(class_to_index).to_numpy()
y_valid_xgb = y_valid_v2.map(class_to_index).to_numpy()

print("\nClasses:")
print(classes)

print("\nNumber of classes:", len(classes))


# ------------------------------------------------------------
# 2. Create XGBoost model
# ------------------------------------------------------------

g2_model = XGBClassifier(
    n_estimators=300,
    max_depth=3,
    learning_rate=0.10,
    subsample=0.8,
    colsample_bytree=0.8,
    objective="multi:softprob",
    num_class=len(classes),
    eval_metric="mlogloss",
    tree_method="hist",
    n_jobs=-1,
    random_state=42,
)


print("\nG2 model:")
print(g2_model)


# ------------------------------------------------------------
# 3. Train
# ------------------------------------------------------------

print("\nTraining G2...")

train_start = time.perf_counter()

g2_model.fit(
    X_train_v2,
    y_train_xgb,
    sample_weight=sample_weights_v2,
)

train_time_g2 = time.perf_counter() - train_start

print(
    f"Training completed in {train_time_g2:.2f} seconds"
)


# ------------------------------------------------------------
# 4. Predict
# ------------------------------------------------------------

print("\nPredicting DEV_VALIDATION...")

prediction_start = time.perf_counter()

y_pred_xgb_index = g2_model.predict(
    X_valid_v2
)

prediction_time_g2 = (
    time.perf_counter() - prediction_start
)


# Convert 0...11 back to lithology codes

y_pred_g2 = np.array([
    index_to_class[int(i)]
    for i in y_pred_xgb_index
])


# ------------------------------------------------------------
# 5. Metrics
# ------------------------------------------------------------

accuracy_g2 = accuracy_score(
    y_valid_v2,
    y_pred_g2
)

balanced_accuracy_g2 = balanced_accuracy_score(
    y_valid_v2,
    y_pred_g2
)

macro_f1_g2 = f1_score(
    y_valid_v2,
    y_pred_g2,
    average="macro",
    zero_division=0,
)

weighted_f1_g2 = f1_score(
    y_valid_v2,
    y_pred_g2,
    average="weighted",
    zero_division=0,
)


# ------------------------------------------------------------
# 6. Results
# ------------------------------------------------------------

print("\n" + "=" * 60)
print("G2 — XGBOOST + GRADIENTS")
print("=" * 60)

print(f"Accuracy          : {accuracy_g2:.6f}")
print(f"Balanced Accuracy : {balanced_accuracy_g2:.6f}")
print(f"Macro F1          : {macro_f1_g2:.6f}")
print(f"Weighted F1       : {weighted_f1_g2:.6f}")

print(f"\nTraining time     : {train_time_g2:.2f} s")
print(f"Prediction time   : {prediction_time_g2:.2f} s")


# ------------------------------------------------------------
# 7. Classification report
# ------------------------------------------------------------

print("\nClassification report:")

g2_report = classification_report(
    y_valid_v2,
    y_pred_g2,
    labels=classes,
    zero_division=0,
)

print(g2_report)

✓ XGBoost is available.

Classes:
[30000 65000 65030 70000 70032 74000 80000 86000 88000 90000 93000 99000]

Number of classes: 12

G2 model:
XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=0.8, device=None, early_stopping_rounds=None,
              enable_categorical=True, eval_metric='mlogloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=3, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=300, n_jobs=-1, num_class=12, ...)

Training G2...
Training completed in 170.42 seconds

Predicting DEV_VALIDATION...

G2 — XGBOOST + GRADIENTS
A

| Metric | G1 ExtraTrees + gradients | G2 XGBoost + gradients | Change |
|---|---:|---:|---:|
| Accuracy | 0.720122 | **0.712692** | ↓ 0.00743 |
| Balanced Accuracy | 0.400922 | **0.409914** | ↑ 0.00899 |
| Macro F1 | 0.302777 | **0.369926** | **↑ 0.06715** |
| Weighted F1 | 0.700995 | **0.708203** | ↑ 0.00721 |
| 65030 F1 | 0.28 | **0.30** | ↑ |
| 65030 Recall | 0.26 | **0.28** | ↑ |


G2 — XGBoost + gradients is currently the better gradient-based development model than G1 ExtraTrees + gradients.

In [31]:
# ============================================================
# CELL 22 — G3: LIGHTGBM + GRADIENTS
# ============================================================

import time
import numpy as np
import lightgbm as lgb

# ------------------------------------------------------------
# 1. Create LightGBM model
# ------------------------------------------------------------

g3_model = lgb.LGBMClassifier(
    objective="multiclass",
    num_class=len(classes),

    n_estimators=300,
    learning_rate=0.05,
    max_depth=-1,
    num_leaves=31,

    subsample=0.8,
    colsample_bytree=0.8,

    reg_lambda=1.0,

    n_jobs=-1,
    random_state=42,

    verbosity=-1,
)

print("G3 model:")
print(g3_model)


# ------------------------------------------------------------
# 2. Train
# ------------------------------------------------------------

print("\nTraining G3...")

train_start = time.perf_counter()

g3_model.fit(
    X_train_v2,
    y_train_xgb,
    sample_weight=sample_weights_v2,
)

train_time_g3 = time.perf_counter() - train_start

print(
    f"Training completed in {train_time_g3:.2f} seconds"
)


# ------------------------------------------------------------
# 3. Predict DEV_VALIDATION
# ------------------------------------------------------------

print("\nPredicting DEV_VALIDATION...")

prediction_start = time.perf_counter()

y_pred_g3_index = g3_model.predict(
    X_valid_v2
)

prediction_time_g3 = (
    time.perf_counter() - prediction_start
)


# Convert encoded class indices back to lithology codes

y_pred_g3 = np.array([
    index_to_class[int(i)]
    for i in y_pred_g3_index
])


# ------------------------------------------------------------
# 4. Calculate metrics
# ------------------------------------------------------------

accuracy_g3 = accuracy_score(
    y_valid_v2,
    y_pred_g3
)

balanced_accuracy_g3 = balanced_accuracy_score(
    y_valid_v2,
    y_pred_g3
)

macro_f1_g3 = f1_score(
    y_valid_v2,
    y_pred_g3,
    average="macro",
    zero_division=0,
)

weighted_f1_g3 = f1_score(
    y_valid_v2,
    y_pred_g3,
    average="weighted",
    zero_division=0,
)


# ------------------------------------------------------------
# 5. Print results
# ------------------------------------------------------------

print("\n" + "=" * 60)
print("G3 — LIGHTGBM + GRADIENTS")
print("=" * 60)

print(f"Accuracy          : {accuracy_g3:.6f}")
print(f"Balanced Accuracy : {balanced_accuracy_g3:.6f}")
print(f"Macro F1          : {macro_f1_g3:.6f}")
print(f"Weighted F1       : {weighted_f1_g3:.6f}")

print(f"\nTraining time     : {train_time_g3:.2f} s")
print(f"Prediction time   : {prediction_time_g3:.2f} s")


# ------------------------------------------------------------
# 6. Complete classification report
# ------------------------------------------------------------

g3_report_df = pd.DataFrame(
    classification_report(
        y_valid_v2,
        y_pred_g3,
        labels=classes,
        output_dict=True,
        zero_division=0,
    )
).T

print("\nClassification report:")

display(g3_report_df)

G3 model:
LGBMClassifier(colsample_bytree=0.8, learning_rate=0.05, n_estimators=300,
               n_jobs=-1, num_class=12, objective='multiclass', random_state=42,
               reg_lambda=1.0, subsample=0.8, verbosity=-1)

Training G3...
Training completed in 163.36 seconds

Predicting DEV_VALIDATION...


/Users/akashray/Lithology_Predictor/.venv/lib/python3.13/site-packages/sklearn/metrics/_classification.py:2939: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")



G3 — LIGHTGBM + GRADIENTS
Accuracy          : 0.713495
Balanced Accuracy : 0.404474
Macro F1          : 0.347732
Weighted F1       : 0.709796

Training time     : 163.36 s
Prediction time   : 31.25 s

Classification report:


,precision,recall,f1-score,support
30000,0.605219,0.729724,0.661665,38013.000000
65000,0.863185,0.830381,0.846465,176301.000000
65030,0.340281,0.342207,0.341241,29339.000000
70000,0.520346,0.696404,0.595638,12431.000000
70032,0.041444,0.012039,0.018658,2575.000000
74000,0.118421,0.036364,0.055641,495.000000
80000,0.360731,0.295880,0.325103,10947.000000
86000,0.675214,0.329167,0.442577,240.000000
88000,1.000000,0.000513,0.001025,3899.000000
90000,0.548611,0.768482,0.640194,514.000000


comparision

| Metric | G1 ExtraTrees | G2 XGBoost | **G3 LightGBM** |
|---|---:|---:|---:|
| Accuracy | 0.720122 | **0.712692** | **0.713495** |
| Balanced Accuracy | 0.400922 | **0.409914** | **0.404474** |
| Macro F1 | 0.302777 | **0.369926** | 0.347732 |
| Weighted F1 | 0.700995 | **0.708203** | 0.709796 |
| Training time | 26.97 s | 170.42 s | 163.36 s |
| Prediction time | 1.53 s | 3.09 s | 31.25 s |

65030 sandstone/shale
| Model | Precision | Recall | F1 |
|---|---:|---:|---:|
| ExtraTrees | .31 | .26 | .28 |
| XGBoost | .33 | .28 | **.30** |
| LightGBM | .34 | .34 | **.34** |

86000 — Anhydrite

| Model | F1 |
|---|---:|
| ExtraTrees | .11 |
| XGBoost | **.41** |
| LightGBM | **.44** |

90000 coal
| Model | F1 |
|---|---:|
| ExtraTrees | .62 |
| XGBoost | .52 |
| LightGBM | **.64** |

Experiment Report --- Gradient Features and Tree-Based Models
1. Experiment Objective
This experiment investigated whether depth-based geological gradient
features improve lithology prediction when added to the existing W5
feature representation.
It also compared three tree-based learners under the same experimental
protocol:
- Extra Trees
- XGBoost
- LightGBM
The purpose was to determine whether gradients are useful, which learner
can exploit them most effectively, and whether difficult
lithologies---especially 65030 (Sandstone/Shale)---improve.
The official PUBLIC and BLIND_TEST wells were not used for model
development.
2. Dataset and Experimental Setup
The engineered dataset contained 1,430,974 labelled samples from 118
wells.
The official FORCE 2020 well split was preserved:
- 98 official TRAIN wells
- 10 PUBLIC wells
- 10 BLIND_TEST wells
Only the 98 official TRAIN wells were used for development.
Those 98 wells were divided using random_state=42 into:
- 78 DEV_TRAIN wells
- 20 DEV_VALIDATION wells
The development split was saved as:
data/splits/development_split.csv
The PUBLIC and BLIND_TEST wells remained untouched.
The split was performed at the well level, not at the
individual-sample level. This is important because nearby depth samples
from the same well are highly correlated. A random row split could
therefore make validation unrealistically easy.
3. Gradient Feature Construction
Five new gradient features were calculated:
- GR_GRAD
- RDEP_LOG10_GRAD
- RMED_LOG10_GRAD
- DTC_GRAD
- RHOB_GRAD
A gradient represents the rate of change of a log with depth:
\(`\frac{dX}{dz}`{=tex}\)
For example:
\(GR_GRAD `\approx `{=tex}`\frac{GR_{i+1}-GR_i}{z_{i+1}-z_i}`{=tex}\)
A positive gradient means the log is increasing with depth, while a
negative gradient means it is decreasing.
The gradients were calculated from the complete LAS depth sequences,
rather than from the already filtered labelled dataset.
The original LAS sequences were checked and showed regular approximately
0.152 m depth spacing:
- 2,337,521 normal transitions
- 0 abnormal transitions
Earlier large gaps in the labelled dataset were therefore mainly
consequences of filtering/removing rows during preprocessing, not gaps
in the original LAS measurements.
4. Why Gradients Can Help
A raw well-log value describes the property at a particular depth.
A gradient adds another piece of information:
How rapidly is that property changing with depth?

Lithological boundaries are often associated with changes in physical
properties. For example:
- GR gradients can indicate changes between cleaner and shalier
  intervals.
- DTC gradients can indicate changes in acoustic properties.
- RHOB gradients can indicate density contrasts.
- Resistivity gradients can indicate changes in formation/fluid
  properties.
Two samples with similar absolute log values can therefore have
different geological contexts if one is inside a stable interval and the
other is near a sharp transition.
Gradient features attempt to provide this local transition information
to the model.
5. Feature Set V2
The gradient features were aligned to the existing engineered dataset
using:
WELL + DEPT
Both datasets had zero duplicate (WELL, DEPT) keys.
The merge preserved exactly 1,430,974 rows.
Feature Set V2 contains 52 total columns, including metadata, but only
46 columns were used as ML features:
- the existing 41 W5 features,
- five gradient features.
DEPT was deliberately excluded from the ML feature matrix.
Depth was retained for gradient calculation and sample alignment, but
absolute depth was not supplied to the model. This avoids encouraging
the model to learn dataset-specific depth shortcuts.
Feature Set V2 was saved as:
data/processed/engineered_features_v2.parquet
No infinite values were present in the final 46-feature matrix.
6. Gradient Missingness
After alignment to the labelled ML dataset, gradient missingness was
approximately:
- GR gradient: 0.092%
- RDEP log gradient: 0.776%
- RMED log gradient: 3.470%
- DTC gradient: 6.056%
- RHOB gradient: 13.144%
RHOB has the highest missingness because RHOB itself has the largest
missingness among the principal model curves.
7. Development Split and Rare Classes
The development validation set contains examples of almost all
lithologies.
The exception is 93000 (Basement).
Basement occurs in only one official TRAIN well. That well was retained
in DEV_TRAIN because moving it to validation would leave the model with
zero Basement examples for training.
Therefore Basement cannot be meaningfully evaluated on this particular
development split.
This is a limitation of the dataset, not a model failure.
It also demonstrates why rare-lithology evaluation must consider
well-level geological diversity, not only the number of individual
samples.
8. Training Weighting
All three experiments used the same square-root class-balancing
strategy.
For class (c):
\frac{N}{K N_c}
}`{=tex} \]

where:

-   \(N\) is the number of DEV_TRAIN samples,
-   \(K\) is the number of classes,
-   (N_c) is the number of DEV_TRAIN samples in class (c).

The weights were calculated **only from DEV_TRAIN**.

The validation data were not used to calculate training weights.

The resulting sample weights ranged approximately from **0.370 to
22.999**.

This gives rare lithologies greater influence during training without
using the validation distribution to construct the model.

------------------------------------------------------------------------

# 9. Experiment G1 --- Extra Trees + Gradients

G1 used the existing W5 Extra Trees architecture with the five new
gradient features.

Configuration:

-   100 trees
-   maximum depth = 20
-   minimum samples per leaf = 5
-   no built-in class weighting
-   external square-root-balanced sample weights
-   random state = 42

Results:

-   Accuracy: **0.720122**
-   Balanced Accuracy: **0.400922**
-   Macro F1: **0.302777**
-   Weighted F1: **0.700995**

For 65030 Sandstone/Shale:

-   Precision: 0.31
-   Recall: 0.26
-   F1: 0.28

Several rare classes remained weak:

-   70032 Chalk: F1 = 0.00
-   74000 Dolomite: F1 = 0.04
-   86000 Anhydrite: F1 = 0.11
-   88000 Halite: F1 = 0.00

Therefore, adding gradients to Extra Trees did not by itself solve the
difficult-class problem.

------------------------------------------------------------------------

# 10. Experiment G2 --- XGBoost + Gradients

G2 kept the same:

-   78 DEV_TRAIN wells,
-   20 DEV_VALIDATION wells,
-   46 features,
-   sample weights,
-   random seed,

and changed only the learner to XGBoost.

Configuration:

-   300 estimators
-   learning rate = 0.10
-   maximum depth = 3
-   subsample = 0.8
-   column subsampling = 0.8
-   multiclass soft-probability objective
-   histogram tree method
-   12 classes
-   random state = 42

Results:

-   Accuracy: **0.712692**
-   Balanced Accuracy: **0.409914**
-   Macro F1: **0.369926**
-   Weighted F1: **0.708203**

For 65030:

-   Precision: 0.33
-   Recall: 0.28
-   F1: **0.30**

Notably, Anhydrite 86000 improved substantially compared with G1:

-   G1 F1 = 0.11
-   G2 F1 = 0.41

Dolomite also improved:

-   G1 F1 = 0.04
-   G2 F1 = 0.09

The improvement in Macro F1 indicates that boosted trees were exploiting
the engineered feature representation differently from Extra Trees.

------------------------------------------------------------------------

# 11. Experiment G3 --- LightGBM + Gradients

G3 used LightGBM while keeping the same features, wells, sample weights,
and random seed.

Configuration:

-   300 estimators
-   learning rate = 0.05
-   31 leaves
-   maximum depth unrestricted
-   subsample = 0.8
-   column subsampling = 0.8
-   L2 regularization = 1.0
-   random state = 42

Results:

-   Accuracy: **0.713495**
-   Balanced Accuracy: **0.404474**
-   Macro F1: **0.347732**
-   Weighted F1: **0.709796**

For 65030:

-   Precision: **0.340281**
-   Recall: **0.342207**
-   F1: **0.341241**

This is important because LightGBM produced the strongest 65030 result
among the three gradient experiments.

Other notable LightGBM results included:

-   86000 Anhydrite F1 = 0.442577
-   90000 Coal F1 = 0.640194
-   99000 Tuff F1 = 0.244579

Halite remained extremely difficult:

-   Precision = 1.000000
-   Recall = 0.000513
-   F1 = 0.001025

This means LightGBM almost never predicted Halite. When it did, it was
correct, but it missed nearly all true Halite samples.

------------------------------------------------------------------------

# 12. Meaning of Accuracy

Accuracy is:

\[ Accuracy = `\frac{\text{correct predictions}}`{=tex}
{`\text{total predictions}`{=tex}} \]

It is useful, but the dataset is strongly imbalanced. Shale 65000
represents a large fraction of the samples.

Therefore a model can have reasonable accuracy while performing poorly
on rare lithologies.

Accuracy should therefore not be the sole selection metric for this
project.

------------------------------------------------------------------------

# 13. Meaning of Balanced Accuracy

Balanced accuracy averages the recall of every class:

\[ Balanced Accuracy = `\frac{1}{K}`{=tex} `\sum`{=tex}\_{c=1}\^{K}
Recall_c \]

Every class therefore receives equal importance.

This is more informative than raw accuracy when rare lithologies matter.

The three gradient experiments produced balanced accuracies around 0.40,
showing some improvement in class-balanced performance but also
substantial remaining confusion.

------------------------------------------------------------------------

# 14. Meaning of F1

For an individual class:

\[ F1 = `\frac{2PR}{P+R}`{=tex} \]

where:

-   \(P\) = precision
-   \(R\) = recall

Precision asks:

> When the model predicts this lithology, how often is it correct?

Recall asks:

> Of all true samples of this lithology, how many did the model find?

F1 balances both.

This is particularly useful for rare lithologies because predicting a
rare class excessively can increase recall while destroying precision.

------------------------------------------------------------------------

# 15. Macro F1

Macro F1 calculates the F1 score separately for every class and gives
every class equal importance:

\[ MacroF1 = `\frac{1}{K}`{=tex} `\sum`{=tex}\_{c=1}\^{K}F1_c \]

This is one of the most important development metrics for this project
because it prevents the dominant Shale class from hiding poor rare-class
performance.

The best Macro F1 in this experiment was produced by XGBoost:

**0.369926**

LightGBM produced:

**0.347732**

Extra Trees produced:

**0.302777**

------------------------------------------------------------------------

# 16. Weighted F1

Weighted F1 weights each class according to its sample count:

\[ WeightedF1 = `\sum`{=tex}\_c `\frac{N_c}{N}`{=tex}F1_c \]

It is therefore strongly influenced by common lithologies.

A high Weighted F1 can coexist with poor rare-class performance.

It should therefore be treated as a secondary metric rather than the
primary rare-class selection metric.

------------------------------------------------------------------------

# 17. Why XGBoost Performed Best Overall

Extra Trees constructs many randomized trees independently and averages
them.

Boosting methods construct trees sequentially, with later trees
attempting to correct errors made by previous trees.

This sequential error-correction mechanism can be useful when lithology
classes overlap strongly in feature space.

The experiment supports the hypothesis that boosted trees can exploit
the combination of:

-   raw log values,
-   rolling statistics,
-   missingness indicators,
-   resistivity separation,
-   depth gradients.

This does not prove that XGBoost will remain the best model after
additional feature engineering, but it makes XGBoost a strong candidate
for later experiments.

------------------------------------------------------------------------

# 18. Why LightGBM Remains Important

Although XGBoost had the highest overall Macro F1, LightGBM performed
better specifically on the main difficult target, 65030:

\[ F1\_{65030}=0.341241 \]

compared with approximately 0.30 for XGBoost.

Therefore LightGBM should not be discarded.

The eventual model selection should consider:

-   Macro F1,
-   Balanced Accuracy,
-   65030 performance,
-   rare-class performance,
-   FORCE penalty score,
-   generalization to unseen wells.

------------------------------------------------------------------------

# 19. Current Scientific Conclusions

### Gradient features are technically valid

They were calculated from the original regular 0.152 m LAS depth
sequences and correctly aligned to the labelled ML dataset.

### Gradients alone do not solve the difficult classes

Adding five gradients to Extra Trees did not produce a decisive
improvement.

65030, Chalk, Dolomite, and Halite remain difficult.

### Boosted trees are promising

Both XGBoost and LightGBM performed better than Extra Trees in Macro F1.

XGBoost achieved the highest overall Macro F1:

**0.369926**

LightGBM achieved the strongest 65030 F1:

**0.341241**

This supports continuing development with boosted-tree models.

------------------------------------------------------------------------

# 20. Important Limitations

These are **development-validation results**, not final blind results.

The 10 official BLIND_TEST wells have not been used for tuning based on
these experiments.

The Basement class cannot be evaluated on this particular development
validation split because no Basement sample is present in
DEV_VALIDATION.

The class-weighting strategy can also produce undesirable
precision/recall trade-offs for extremely rare classes. For example,
high recall can occur alongside very low precision.

Future experiments should therefore examine both class weighting and
probability/decision behavior.

------------------------------------------------------------------------

# 21. Next Experiment

The next notebook will investigate **finer vertical rolling/context
features**.

Current windows are:

-   5 samples ≈ 0.760 m
-   9 samples ≈ 1.368 m
-   21 samples ≈ 3.192 m

The next experiment will investigate:

-   3 samples ≈ 0.456 m
-   5 samples ≈ 0.760 m
-   7 samples ≈ 1.064 m
-   9 samples ≈ 1.368 m
-   15 samples ≈ 2.280 m
-   21 samples ≈ 3.192 m

The motivation is that lithological structures occur at different
vertical scales.

Small windows can capture sharp boundaries and thin beds.

Intermediate windows can capture local lithological packages.

Larger windows can capture broader formation-scale context.

The development well split will remain frozen.

The official PUBLIC and BLIND_TEST wells will remain untouched.

------------------------------------------------------------------------

# 22. Experiment Status

Feature Set V2 is complete and reproducible:

`data/processed/engineered_features_v2.parquet`

Development split:

`data/splits/development_split.csv`

Current best overall gradient experiment:

**XGBoost + gradients**

Development-validation metrics:

-   Accuracy: **0.712692**
-   Balanced Accuracy: **0.409914**
-   Macro F1: **0.369926**
-   Weighted F1: **0.708203**
-   65030 F1: **0.30**

Current best 65030 gradient experiment:

**LightGBM + gradients**

-   65030 Precision: **0.340281**
-   65030 Recall: **0.342207**
-   65030 F1: **0.341241**

The official blind dataset remains locked.